routing/per_route_options.go

```diff
@@ -0,0 +1,129 @@
+
+import (
+	"fmt"
+	"path/filepath"
+	"regexp"
+	"slices"
+	"sync"
+	"time"
+
+	"github.com/cloudfoundry/cf-test-helpers/v2/cf"
+	"github.com/cloudfoundry/cf-test-helpers/v2/helpers"
+
+	. "github.com/cloudfoundry/cf-acceptance-tests/cats_suite_helpers"
+	"github.com/cloudfoundry/cf-acceptance-tests/helpers/app_helpers"
+	"github.com/cloudfoundry/cf-acceptance-tests/helpers/assets"
+	"github.com/cloudfoundry/cf-acceptance-tests/helpers/random_name"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	. "github.com/onsi/gomega/gexec"
+)
+
+var _ = RoutingDescribe("Per-Route Options", func() {
+	var (
+		appName        string
+		appId          string
+		instanceIds    [2]string
+		leastConnHost  string
+		roundRobinHost string
+	)
+
+	Context("when an app sets the loadbalancing algorithm", func() {
+		BeforeEach(func() {
+			appName = random_name.CATSRandomName("APP")
+			asset := assets.NewAssets()
+			leastConnHost = random_name.CATSRandomName("dora-lc")
+			roundRobinHost = random_name.CATSRandomName("dora-rr")
+			Expect(cf.Cf("push",
+				appName,
+				"-b", Config.GetRubyBuildpackName(),
+				"-m", DEFAULT_MEMORY_LIMIT,
+				"-p", asset.Dora,
+				"--var", fmt.Sprintf("domain=%s", Config.GetAppsDomain()),
+				"--var", fmt.Sprintf("name=%s", appName),
+				"--var", fmt.Sprintf("leastconnhost=%s", leastConnHost),
+				"--var", fmt.Sprintf("roundrobinhost=%s", roundRobinHost),
+				"-f", filepath.Join(asset.Dora, "man_route_opt_lb.yml"),
+			).Wait(Config.CfPushTimeoutDuration())).To(Exit(0))
+			appId = app_helpers.GetAppGuid(appName)
+			appInstanceRegex, _ := regexp.Compile("[[:alnum:]]{8}(-[[:alnum:]]{4}){4}")
+			for i := range 2 {
+				for {
+					fmt.Fprintf(GinkgoWriter, "Waiting for app instance %d to start...\n", i)
+					curl := helpers.Curl(Config, Config.Protocol()+leastConnHost+"."+Config.GetAppsDomain()+"/id", "-H", fmt.Sprintf("X-Cf-App-Instance: %s:%d", appId, i)).Wait()
+					id := string(curl.Out.Contents())
+					if appInstanceRegex.MatchString(id) {
+						instanceIds[i] = id
+						fmt.Fprintf(GinkgoWriter, "App instance %d has started. Instance ID: %s.\n", i, id)
+						break
+					}
+					fmt.Fprintf(GinkgoWriter, "App instance %d is not ready yet. Response: %s, curl error: %s.\n", i, id, string(curl.Err.Contents()))
+					time.Sleep(1 * time.Second)
+				}
+			}
+		})
+
+		AfterEach(func() {
+			app_helpers.AppReport(appName)
+			Expect(cf.Cf("delete", appName, "-f", "-r").Wait()).To(Exit(0))
+		})
+
+		Context("when it's set to round-robin", func() {
+			It("distributes requests evenly", func() {
+				doraUrl := fmt.Sprintf("%s%s.%s", Config.Protocol(), roundRobinHost, Config.GetAppsDomain())
+				var wg sync.WaitGroup
+				for i := 0; i < 10; i++ {
+					wg.Add(1)
+					go func() {
+						defer wg.Done()
+						defer GinkgoRecover()
+						Expect(helpers.Curl(Config, fmt.Sprintf("%s/delay/20", doraUrl), "-H", fmt.Sprintf("X-Cf-App-Instance: %s:0", appId)).Wait(25 * time.Second)).To(Exit(0))
```

Comment:
This expectation will likely behave strangely due to the asynchronous nature. Is there some other mechanism you can use here instead?

Reply:
We could drop the expect and ignore the return value from the cURL, e.g.:

```suggestion
						helpers.Curl(Config, fmt.Sprintf("%s/delay/20", doraUrl), "-H", fmt.Sprintf("X-Cf-App-Instance: %s:0", appId))
```

Not sure what else we could do here as we need those requests to stay active and cause the imbalance between instances needed to check the algorithm used by gorouter.